engines/bladerunner/subtitles.cpp

```diff
@@ -0,0 +1,778 @@
+*/
+bool Subtitles::hide() {
+	if (!_isVisible) {
+		return false;
+	}
+
+	_isVisible = false;
+	return true;
+}
+
+/**
+* Checks whether the subtitles should be visible or not
+* @return the value of the _isVisible member boolean var
+*/
+bool Subtitles::isVisible() const {
+	return _isVisible;
+}
+
+/**
+* Tick method specific for outtakes (VQA videos)
+*/
+void Subtitles::tickOuttakes(Graphics::Surface &s) {
+	if (_currentSubtitleTextFull.empty()) {
+		_vm->_subtitles->hide();
+	} else {
+		_vm->_subtitles->show();
+	}
+#if BLADERUNNER_RESTORED_CONTENT_GAME
+	if (!_vm->_extraGameFlagsForRestoredContent->query(kEDSFlagSubtitlesEnable)) {
+		return;
+	}
+#endif
+	if (!_isVisible) { // keep it as a separate if
+		return;
+	}
+	draw(s);
+}
+
+/**
+* Tick method for in-game subtitles -- Not for outtake cutscenes (VQA videos)
+*/
+void Subtitles::tick(Graphics::Surface &s) {
+	if (!_vm->_audioSpeech->isPlaying()) {
+		_vm->_subtitles->hide(); // TODO might need a better system. Don't call it always.
+
+	}
+#if BLADERUNNER_RESTORED_CONTENT_GAME
+	if (!_vm->_extraGameFlagsForRestoredContent->query(kEDSFlagSubtitlesEnable)) {
+		return;
+	}
+#endif
+	if (!_isVisible)  { // keep it as a separate if
+		return;
+	}
+	draw(s);
+}
+
+/**
+* Draw method for drawing the subtitles on the display surface
+*/
+void Subtitles::draw(Graphics::Surface &s) {
+	if (!_isVisible || _currentSubtitleTextFull.empty() || !_subsFontsLoaded) {
+		return;
+	}
+	if (_subtitlesQuoteChanged) {
+		calculatePosition(); // Don't always call calc position, only when quote has changed
+		_subtitlesQuoteChanged = false;
+	}
+
+#if SUBTITLES_EXTERNAL_FONT
+	for (int i = 0; i < _currentSubtitleLines; ++i) {
+		_subsFont->draw(_subtitleLineQuote[i], s, _subtitleLineScreenX[i], _subtitleLineScreenY[i]);
+	}
+#else
+	// INTERNAL FONT. NEEDS HACK (_subsBgFont) FOR SHADOW EFFECT
+	for (int i = 0; i < _currentSubtitleLines; ++i) {
+		_subsBgFont->draw(_subtitleLineQuote[i], s, _subtitleLineScreenX[i], _subtitleLineScreenY[i] - 1);
+		_subsBgFont->draw(_subtitleLineQuote[i], s, _subtitleLineScreenX[i], _subtitleLineScreenY[i] + 1);
+		_subsBgFont->draw(_subtitleLineQuote[i], s, _subtitleLineScreenX[i] + 1, _subtitleLineScreenY[i] + 1);
+		_subsBgFont->draw(_subtitleLineQuote[i], s, _subtitleLineScreenX[i] + 1, _subtitleLineScreenY[i] - 1);
+		if (_subtitleLineScreenX[i] > 0) {
+			_subsBgFont->draw(_subtitleLineQuote[i], s, _subtitleLineScreenX[i] - 1, _subtitleLineScreenY[i] - 1);
+			_subsBgFont->draw(_subtitleLineQuote[i], s, _subtitleLineScreenX[i] - 1, _subtitleLineScreenY[i] + 1);
+		}
+		_subsFont->draw(_subtitleLineQuote[i], s, _subtitleLineScreenX[i],  _subtitleLineScreenY[i]);
+	}
+#endif // SUBTITLES_EXTERNAL_FONT
+}
+
+/**
+* Calculate the position (X axis - horizontal) where the current active subtitle text should be displayed/drawn
+* This also determines if more than one lines should be drawn and what text goes into each line; splitting into multiple lines is done here
```

Comment:
Why not calculate the bounding box and then just use wordWrap() function plus aligned text rendering here? This could simplify things big time.

Reply:
I wasn't aware of the wordWrap method. Still, the purpose of this code here is to allow subtitle quotes to be split in multiple lines first by checking for new line characters, and then by doing the wrapping thing. However, the wrapping code also makes sure that the line segments are approximately even (or close to that) with regard to their width. Is this something I can do with wordWrap()?